### PR TITLE
Go modules stack submission

### PR DESCRIPTION
**Pull request by jpapejr**
### Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stack-structure.md).

### Modifying an existing stack:

n/a

### Contributing a new stack:

- Describe how application dependencies are managed:

This stack uses Go v1.11 format Modules for dependencies

- Explain how Appsody file watcher is utilized:

File watcher watches .go, .mod files.

- Describe other Appsody environment variables defined by the stack image:

Nothing other than the minimum that Appsody needs.

- Describe any limitations and known issues:

`appsody debug` works despite seeing spurious messages like `could not launch process: fork/exec /project/user-app/__debug_bin: inappropriate ioctl for device`. Running with `appsody debug -I` helps in some cases. 

### Related Issues:
Originally submitted under PR #425 but was split out for clarity

### CODE REVIEW

**Comment by jpapejr** [comment]
Original commits were done with `root` - updated to use proper author info.

**Comment by jpapejr** [comment]
@Kamran64 is that root commit going to bork the CLA check even though I've signed it?

**Comment by Kamran64** [comment]
@jpapejr cheers for splitting these. Looks like it. Where did this root commit come from?

**Comment by jpapejr** [comment]
@Kamran64 From my dev VM.. After I deleted and re-forked your repo I forgot to `git config` my name and email prior to the commit and push. I reset the author and re-pushed which are the subsequent commits you see..

**Comment by Kamran64** [comment]
Hmm - Is there any way you can reauthor commits? I'm not sure if you can... I'm also quite curious as to why the travis build is failing. Tried restarting it but the same error appears.. I'll investigate.

**Comment by jpapejr** [comment]
@Kamran64 I think I was able to go back a rebase the `root` commit and update the author.. I've pushed everything back up and it seems to have repaired things. I'll watch the build.

**Comment by jpapejr** [comment]
Added note to readme here as well about debugging being unsupported at this time.

File: experimental/go-modules/image/project/.dockerignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+**/.git
+**/.gitignore
+**/.vscode
```
Comment (sam-github) [inline experimental/go-modules/image/project/.dockerignore]: missing newline?

File: experimental/go-modules/image/project/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,15 @@
+# Install the app dependencies in a full Node docker image
+FROM golang:alpine
```
Comment (sam-github) [inline experimental/go-modules/image/project/Dockerfile]: README said its go 1.12, should this be `golang:1.12-alpine3.10`?

File: experimental/go-modules/image/project/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+WORKDIR "/project/user-app"
+RUN apk add --update git bash
+RUN go get -u github.com/golang/dep/cmd/dep
+COPY ./user-app/* ./
+RUN ls -ltra /project/user-app
+RUN ls -ltra /go
+RUN if [ -e /project/user-app/.vendor-me.sh ]; then bash /project/user-app/.vendor-me.sh; fi
+RUN go build -o /app ./... 
+RUN chmod +x /app
+CMD ["/app"]
```
Comment (sam-github) [inline experimental/go-modules/image/project/Dockerfile]: missing newline

File: experimental/go-modules/image/project/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,15 @@
+# Install the app dependencies in a full Node docker image
+FROM golang:alpine
+
+
+# Install user-app dependencies
+WORKDIR "/project/user-app"
+RUN apk add --update git bash
+RUN go get -u github.com/golang/dep/cmd/dep
+COPY ./user-app/* ./
+RUN ls -ltra /project/user-app
```
Comment (sam-github) [inline experimental/go-modules/image/project/Dockerfile]: debug code that got left in?

File: experimental/go-modules/image/project/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,15 @@
+# Install the app dependencies in a full Node docker image
+FROM golang:alpine
+
+
+# Install user-app dependencies
+WORKDIR "/project/user-app"
+RUN apk add --update git bash
+RUN go get -u github.com/golang/dep/cmd/dep
```
Comment (sam-github) [inline experimental/go-modules/image/project/Dockerfile]: what's dep for, isn't that an older alternative to go modules?

File: experimental/go-modules/image/project/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+COPY ./user-app/* ./
```
Comment (sam-github) [inline experimental/go-modules/image/project/Dockerfile]: some go projects speed up the build by cacheing the dep layer seperately from the app:
```
COPY go.* ./
RUN go mod download
COPY . .
RUN go build // etc.
```

File: experimental/go-modules/image/Dockerfile-stack [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+COPY ./config /config
+
+WORKDIR /project/user-app
+
+RUN go get -u github.com/go-delve/delve/cmd/dlv
+
+ENV PORT=3000
+
+EXPOSE 3000
+EXPOSE 9229
```
Comment (sam-github) [inline experimental/go-modules/image/Dockerfile-stack]: is this the dlv interface?

File: experimental/go-modules/templates/default/go.mod [hunk 8]
```diff
@@ -0,0 +1 @@
+module github.com/example
```
Comment (sam-github) [inline experimental/go-modules/templates/default/go.mod]: Hopefully obvious to most go developers, but maybe a comment saying that this example value should be changed would be helpful?

File: experimental/go-modules/image/project/Dockerfile (hunk 5 above)
Comment (sam-github) [reply experimental/go-modules/image/project/Dockerfile]: I think this line is a leftover from https://github.com/appsody/stacks/pull/435

**Comment by jpapejr** [comment]
I won’t be making any more updates to this PR. Feel free to close it or merge it. Either way, I don’t have the cycles for more back and forth.

**Comment by Kamran64** [comment]
@kylegc @chilanti @neeraj-laad - I've made some changes here and done more testing. I think this stack is in a better shape to go into experimental right now, over the go-dep stack. Please take a look.

File: experimental/go-modules/image/Dockerfile-stack [hunk 9]
```diff
@@ -0,0 +1,27 @@
+FROM golang:1
```
Comment (kylegc) [inline experimental/go-modules/image/Dockerfile-stack]: The README says this is based on `Go v1.12.5` however this image tag currently pulls `1.13.3`. 
I think you should at least lock this image to `golang:1.12`

File: experimental/go-modules/image/Dockerfile-stack [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+ENV APPSODY_WATCH_DIR=/project/user-app
+ENV APPSODY_WATCH_IGNORE_DIR=/project/user-app/vendor
+ENV APPSODY_WATCH_REGEX="^.*.(go|mod)$"
+
+ENV APPSODY_RUN="go run ."
+ENV APPSODY_RUN_ON_CHANGE="go run ."
+ENV APPSODY_RUN_KILL=false
+
+
+ENV APPSODY_TEST="go test ."
```
Comment (kylegc) [inline experimental/go-modules/image/Dockerfile-stack]: I think this should be `go test ./...`

**Pull request merged**